contracts/gateway/src/contract/execute.rs

```diff
@@ -102,3 +72,49 @@ pub fn route_outgoing_messages(
             .map(|msg| GatewayEvent::MessageRouted { msg }.into()),
     ))
 }
+
+fn partition_by_verified(
+    deps: DepsMut,
+    msgs: Vec<NewMessage>,
+) -> error_stack::Result<(Vec<NewMessage>, Vec<NewMessage>), ContractError> {
+    let verifier = load_config(&deps)?.verifier;
+
+    let query_msg = aggregate_verifier::msg::QueryMsg::IsVerified {
+        messages: msgs.clone(),
+    };
+    let query_response: Vec<(CrossChainId, bool)> = deps
+        .querier
+        .query(&QueryRequest::Wasm(WasmQuery::Smart {
+            contract_addr: verifier.to_string(),
+            msg: to_binary(&query_msg).map_err(ContractError::from)?,
+        }))
+        .map_err(ContractError::from)?;
+
+    Ok(msgs.into_iter().partition(|msg| -> bool {
+        match query_response.iter().find(|r| msg.cc_id == r.0) {
+            Some((_, v)) => *v,
+            None => false,
+        }
+    }))
+}
+
+fn load_config(deps: &DepsMut) -> error_stack::Result<Config, ContractError> {
+    let cfg = CONFIG
+        .load(deps.storage)
+        .change_context(ContractError::ConfigNotFound)?;
+    Ok(cfg)
+}
+
+fn ensure_unique_ids(msgs: &[NewMessage]) -> error_stack::Result<(), ContractError> {
+    let duplicates: Vec<_> = msgs
+        .iter()
+        .map(|m| &m.cc_id)
+        .duplicates()
+        .map(|cc_id| cc_id.to_string())
```

Comment:
```suggestion
        .map(|m| m.cc_id.to_string())
        .duplicates()
```

Reply:
I forgot to add a comment here, there is a reason why I split the maps: I want the duplicates check to use the strongly typed ID instead of the string